---
title: "RNA-Seq Workshop"
authors: "Meagan Archer and Stephanie Ali Fairbairn"
date: "29/03/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
Welcome to workshop portion of today's tutorial!

This is part 1 of 3 sections towards integrating our data: RNA-seq analysis

For more context and details on each step look under the slides folder in the main repo. 
___________________________

The raw reads have been previously quality checked, cleaned and aligned to the reference genome for *Streptomyces venezuelae* by a past member of the Elliot Lab. The next steps in the RNA-sequencing differential analysis pipeline are to:

1. Count the reads mapping to genomic features. 
2. Use the counts generated to test for differentially expressed genes in the mutant (T3d3225) compared to the wildtype (T3WT). 

We will be using the `featureCounts()` function of `RSubread`, a package within the `Bioconductor`
project package repository, to first assign the mapped sequencing reads to the genomic features contained in the *S. venezuelae* genome annotation (.gff) file. We will then use the counts generated as input for the `DESeq2` package, to test for differentially expressed genes. The workflow below assumes that the necessary sorted .bam (containing the aligned reads) and genome annotation (.gff) files have been downloaded and saved to your working directory.

First, install the `RSubread` and `ape` packages. The `ape` package is needed to read in the .gff file. Please see here for the complete documentation for the [RSubread](https://bioconductor.org/packages/release/bioc/vignettes/Rsubread/inst/doc/SubreadUsersGuide.pdf) and 
[ape](https://cran.r-project.org/web/packages/ape/ape.pdf) packages.

```{r}
# Package Install ----

if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("Rsubread")
BiocManager::install("DESeq2")

install.packages("ape")
```

Next, set your working directory using the `setwd()` function. Check that you are in the correct working directory using the `getwd()` function.

```{r}
# Set up working directory ----

setwd("~") # <- Edit this based on the directory of your choice
getwd() # <- Check that you are in the correct working directory
```

Load associated libraries. 

```{r}
# Library ----

library(Rsubread)
library(ape)
library(DESeq2)
library(stringr)
```

###Assign mapped reads to genomic features using featureCounts()

Once you have set up and are currently in your working directory, read the genome annotation (.gff) file into the environment as a data frame using the `read.gff()` function of the `ape` package and save the data frame into the variable, "s_venezuelae". This will allow you to check the structure of the genome annotation file.

Pass the following arguments, `file`, `na.strings` and `GFF3` to `read.gff()`  as: `"s_venezuelae.gff"`,
`c(".", "?")` and `TRUE`, respectively. The `file` argument specifies the annotation file. The `na.strings` argument specifies which characters will be turned into "NAs" in the final format. The ```GFF3``` argument is a logical that specifies whether the annotation file is in .gff3 (```TRUE```) or .gff2 (```FALSE```) format. Perform a "quick check" by printing the first six rows of the data frame using the ```head()``` function. 

```{r, `echo = FALSE`}
# Check the structure of the annotation file

s_venezuelae <- read.gff("s_venezuelae.gff", na.strings = c(".", "?"), GFF3 = TRUE) # <- Read the genome annotation file into the environment as a data frame

head(s_venezuelae) # <- Read the first six rows of the "s_venezuelae" data frame

unique(s_venezuelae$type) # <- Check the data in the "type" column of the "s_venezuelae" data frame
```



Depending on the annotation file, the structure of the data may differ slightly from the default parameters for ```featureCounts()```. ```featureCounts()``` expects the ```GTF.featureType``` to be ```"exon"```  and ```GTF.attrType``` to be ```"gene_id"``` by default.```GTF.featureType``` denotes the feature type (in this case ```type```) used to select the rows of the genome annotation file for read summarization. ```GTF.attrType``` denotes the attributes that will be used to group features into meta-features. 

By examining the structure of the "s_venezuelae" data frame, we can see that the ```GTF.attrType``` corresponds to ```"ID"```. This will need to be specified to ```featureCounts()``` in place of ```"gene_id"```.
  
  
Example Output ```head(s_venezuelae)```
```
seqid  source type start  end score strand phase
1 FR845719 GenBank gene    67  657    NA      -     1
2 FR845719 GenBank gene   166  396    NA      +     1
3 FR845719 GenBank gene   478  696    NA      +     1
4 FR845719 GenBank gene  1064 1678    NA      +     1
5 FR845719 GenBank gene  1987 2235    NA      +     1
6 FR845719 GenBank gene  2368 2784    NA      +     1
                   attributes
1 ID=SVEN_0001;Name=SVEN_0001
2 ID=SVEN_0002;Name=SVEN_0002
3 ID=SVEN_0003;Name=SVEN_0003
4 ID=SVEN_0004;Name=SVEN_0004
5 ID=SVEN_0005;Name=SVEN_0005
6 ID=SVEN_0006;Name=SVEN_0006
```

Since the ```GTF.featureType``` data in the annotation may differ, it is important to examine its corresponding column in the data frame. Use the ```unique()``` function to search the data frame "s_venezuelae", by passing it as the first argument to the function ```unique()```. This will omit any duplicate rows in the ```type``` column, as specified by ```$```. 

```r
# Check the data in the "type" column of the "s_venezuelae" data frame unique(s_venezuelae$type)
```
In this case, the only ```type``` data corresponds to ```gene``` and therefore there is no ```exon``` data. Therefore, ```GTF.featureType``` must be specified as ```gene```
to the ```featureCounts()``` function.

Example Output ```unique(s_venezuelae$type)```
```
[1] gene
Levels: gene
```
We will be counting the reads that are mapped to the genomic features of *S. venezuelae* for four previously sorted .bam files, with two sorted .bam files per genotype (mutant and wildtype), corresponding to two
biological replicates for each. First, specify the ```files``` argument to ```featureCounts()```. The ```files``` argument specifies the input files as a character vector. Aggregate the four sorted .bam files using the ```dir()``` function. ```dir()``` specifies all files in a directory. To specify the current working directory, use, ```"."```, followed by ```"bam$"```, to select only the files ending in .bam. Next, specify the annotation file to the ```annot.ext``` argument as ```"s_venezuelae.gff"```. Set the ```isGTFAnnotationFile``` to ```TRUE```, ```GTF.featureType``` to ```"gene"``` and ```GTF.attrType``` to ```"ID"```. Since we are not If you are interested in summarizing the reads according to meta-features, set ```useMetaFeatures``` to ```FALSE```.

Check the first six rows of the outputed "sorted_counts" using ```head(sorted_counts)```.  
    

```{r, `echo = FALSE`}
# Use featureCounts to count the reads that are mapped to the genomic features 

sorted_counts <- featureCounts(files=dir(".", "bam$"), annot.ext="s_venezuelae.gff", isGTFAnnotationFile=TRUE, GTF.featureType="gene", GTF.attrType="ID", useMetaFeatures = FALSE) 

head(sorted_counts) # <- Check the structure of "sorted_counts" and read the first six rows.

```

###Use DESeq2 to perform differential gene expression analysis

##Count Matrix Input
```{r, `echo = FALSE`}
# Prepare the counts matrix for input

counts <- sorted_counts$counts # <- Subset the counts data from "sorted_counts"

head(counts) # <- Check the first 6 rows of the new "counts" table
```


```{r, `echo = FALSE`}
# Prepare "colData"

files <- c("T3d3225v1.sorted.bam", "T3d3225v2.sorted.bam", "T3WTv1.sorted.bam", "T3WTv2.sorted.bam") # <- Create a vector called, "files", containing the names of the .bam files (will be column one)

genotype <- c("mutant", "mutant", "wildtype", "wildtype") # <-  Create a vector called, "genotype", containing the corresponding genotypes of the above .bam files (will be column two) 

coldata <- data.frame(files, genotype) # <- Create the data frame "coldata"
```


```{r, `echo = FALSE`}
# Generate the DESeqDataSet for comparing differentially expressed genes in the mutant and wildtype

deseqdata <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~genotype)
```


```{r, `echo = FALSE`}
# Generate the DESeqDataSet for comparing differentially expressed genes in the mutant and wildtype

deseqdata <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~genotype)
```


##Differential Gene Expression Analysis
```{r}
# Relevel the factors in "genotype" so that wildtype is the reference level for analysis

deseqdata$genotype <- factor(deseqdata$genotype, levels = c("wildtype","mutant"))
```


```{r, `echo = FALSE`}
# Perform DeSeq() analysis and output to results table

deseqdata <- DESeq(deseqdata)

results <- results(deseqdata, alpha = 0.05) # <- Generate the results table based on the DESeq2 analysis with the alpha factor changed to 0.05

results # <- Check results table in the console

```


```{r, `echo = FALSE`}
# Reorganize the results table to display only significantly expressed genes which exhibit >1 log2fold change

results_lfc <- subset(results, abs(log2FoldChange) > 1) # <- Subset the results table using the subset() function according to an absolute log2FoldChange greater than 1 and saves it as a new table called, results_lfc

resultsOrdered <- results_lfc[order(results_lfc$padj),] # <- Reorder the rows of the results_lfc table  using the order() function, and the column specified using "$", ie) the adjusted p-values, and saves it as a new table called resultsOrdered

resultsOrdered <- subset(resultsOrdered, padj<.05) # <- Subset the reordered table of results accorded to include only adjusted p-values less than 0.05 using the subset() function and save it again into resultsOrdered

summary(resultsOrdered) # <- Check the summary of resultsOrdered 

resultsOrdered[is.na(resultsOrdered$padj), ] # <- Select the adjusted p-values of the resultsOrdered table and pass it to the function is.na() to check the adjusted p-value column for "NA" values. Output should contain: "DataFrame with 0 rows and 6 columns"


write.csv(as.data.frame(resultsOrdered), 
          file="mutant_versus_wildtype_gene_expression_data.csv") # <- Write the reordered results to a .csv file
```


### Identify significantly differentially expressed genes that correlate CHIP-sequencing binding sites to genes that possess altered expression when the Lsr2 gene is deleted


```{r}
# Merge the differential gene expression output with CHIP-Seq output

rnaseq_csv <- read.csv("mutant_versus_wildtype_gene_expression_data.csv") # <- Read in the .csv file

colnames(rnaseq_csv) # <- Check the column names of the .csv file

rnaseq_csv <- rnaseq_csv %>% mutate_all(str_replace_all, "SVEN_", "") # <- Remove the prefix, "SVEN_", in the .csv file by piping the mutate_all() function, with the argument str_replace_all with arguments to replace the specified string, "SVEN_", with blank space " "

head(rnaseq_csv) # <- Check that "SVEN_" was removed

rnaseq_csv <- rnaseq_csv %>% rename(ID = X) # <- Change the name of the column containing the gene IDs to "ID" using the rename() function piped to the .csv file

write.csv(as.data.frame(rnaseq_csv), file = "rnaseq.csv", row.names = FALSE) # <- Write the new .csv file

ChIPseq_csv <- read.csv("ChIP_table_clusters.csv") # <- Read in the Chip-Seq .csv file

merged_seq_data <- merge(rnaseq_csv, ChIPseq_csv, by = c('ID', 'ID')) # <- Merge the CHIP-Seq .csv to the rnaseq_csv using the "ID" column

write.csv(as.data.frame(merged_seq_data), file = "merged_seq_data.csv", row.names = FALSE) # <- Write the merged data to a new .csv file using row.names = FALSE to prevent the addition of a column of numbered rows
```
