---
title: "RNA-Seq Workshop"
authors: "Meagan Archer and Stephanie Ali Fairbairn"
date: "29/03/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
# Package Install ----
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("Rsubread")

install.packages("ape")
```


```{r}
# Set up working directory ----
setwd("~") # <- Edit this based on the directory of your choice
getwd() # <- Check that you are in the correct working directory
```


```{r}
# Library ----
library(Rsubread)

library(ape)

library(DESeq2)
```

###Assign mapped reads to genomic features using featureCounts()

```{r, `echo = FALSE`}
# Check the structure of the annotation file

s_venezuelae <- read.gff("s_venezuelae.gff", na.strings = c(".", "?"), GFF3 = TRUE) # <- Read the genome annotation file into the environment as a data frame

head(s_venezuelae) # <- Read the first six rows of the "s_venezuelae" data frame

unique(s_venezuelae$type) # <- Check the data in the "type" column of the "s_venezuelae" data frame
```


```{r, `echo = FALSE`}
# Use featureCounts to count the reads that are mapped to the genomic features 

sorted_counts <- featureCounts(files=dir(".", "bam$"), annot.ext="s_venezuelae.gff", isGTFAnnotationFile=TRUE, GTF.featureType="gene", GTF.attrType="ID", useMetaFeatures = FALSE) 

head(sorted_counts) # <- Check the structure of "sorted_counts" and read the first six rows.

```

###Use DESeq2 to perform differential gene expression analysis

##Count Matrix Input
```{r, `echo = FALSE`}
# Prepare the counts matrix for input

counts <- sorted_counts$counts # <- Subset the counts data from "sorted_counts"

head(counts) # <- Check the first 6 rows of the new "counts" table
```


```{r, `echo = FALSE`}
# Prepare "colData"

files <- c("T3d3225v1.sorted.bam", "T3d3225v2.sorted.bam", "T3WTv1.sorted.bam", "T3WTv2.sorted.bam") # <- Create a vector called, "files", containing the names of the .bam files (will be column one)

genotype <- c("mutant", "mutant", "wildtype", "wildtype") # <-  Create a vector called, "genotype", containing the corresponding genotypes of the above .bam files (will be column two) 

coldata <- data.frame(files, genotype) # <- Create the data frame "coldata"
```



```{r, `echo = FALSE`}
# Generate the DESeqDataSet for comparing deferentially expressed genes in the mutant and wildtype

deseqdata <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~genotype)
```


```{r, `echo = FALSE`}
# Generate the DESeqDataSet for comparing deferentially expressed genes in the mutant and wildtype

deseqdata <- DESeqDataSetFromMatrix(countData = counts, colData = coldata, design = ~genotype)
```

##Differential Gene Expression Analysis
```{r}
# Relevel the factors in "genotype" so that wildtype is the reference level for analysis
deseqdata$genotype <- factor(deseqdata$genotype, levels = c("wildtype","mutant"))
```


```{r, `echo = FALSE`}
# Perform DeSeq() analysis and output to results table
deseqdata <- DESeq(deseqdata)

results <- results(deseqdata) # <- Generate the results table

results # <- View results table in the console
```


```{r, `echo = FALSE`}
# Reorders the results table with lowest adjusted p-value first
resultsOrdered <- results[order(results$padj),]

resultsOrdered # <- View reordered results in the console
```

```{r}
# Write reordered results to a .csv file
write.csv(as.data.frame(resultsOrdered), 
          file="mutant_versus_wildtype_gene_expression_data.csv")
```


