# Data_Integration_Exercise


Initial Outline:

Biol722 group workshop 

Meagan Archer 
Meghan Pepler
Sreedevi Kesavan 
Stephanie Ali Fairbairn 

Integrating RNA-seq with ChIP-Seq 

Introduction to our model system (probably data from the Elliot lab - comparing WT to mutant strain) 
Brief overview of our RNA-seq pipeline 
We have counts tables - now what? 
Narrowing down differentially expressed genes of interest
Introduction to ChIP-seq workflow 
Processing ChIP-Seq data 
Integrating RNA-seq data with ChIP-seq (we are probably going to focus on a transcription factor in this workshop, but the same process can be used to analyze ChIP data for chromatin regulators) 
To predict whether the transcription factor has activating or repressive function
To infer the transcription factor's target genes (direct vs indirect effects of transcription factor on differentially expressed genes)
Reduce false-positive rate
Possibly (time allowing?) find sequence motifs of binding sites near upregulated or downregulated targets to identify collaborating factors
Statistical analyses necessary to gain meaningful information from the integrated data 

** We initially considered mapping transcription start sites, which can be used to identify novel transcripts (possibly under the regulation of the transcription factor we investigate with ChIP-seq). We decided that these additional analyses were pretty involved and thought it might be better to only focus on integrating ChIP-seq and RNA-seq so that we can go more in depth with those analyses. 







